src/Servers/Kestrel/Core/src/Internal/Http/Http1ChunkedEncodingMessageBody.cs

```diff
@@ -0,0 +1,545 @@
+        {
+            if (readableBuffer.IsSingleSegment)
+            {
+                writableBuffer.Write(readableBuffer.First.Span);
+            }
+            else
+            {
+                foreach (var memory in readableBuffer)
+                {
+                    writableBuffer.Write(memory.Span);
+                }
+            }
+        }
+
+        protected override void OnReadStarted()
+        {
+            _pumpTask = PumpAsync();
+        }
+
+        protected bool Read(ReadOnlySequence<byte> readableBuffer, PipeWriter writableBuffer, out SequencePosition consumed, out SequencePosition examined)
+        {
+            consumed = default;
+            examined = default;
+
+            while (_mode < Mode.Trailer)
+            {
+                if (_mode == Mode.Prefix)
+                {
+                    ParseChunkedPrefix(readableBuffer, out consumed, out examined);
+
+                    if (_mode == Mode.Prefix)
+                    {
+                        return false;
+                    }
+
+                    readableBuffer = readableBuffer.Slice(consumed);
+                }
+
+                if (_mode == Mode.Extension)
+                {
+                    ParseExtension(readableBuffer, out consumed, out examined);
+
+                    if (_mode == Mode.Extension)
+                    {
+                        return false;
+                    }
+
+                    readableBuffer = readableBuffer.Slice(consumed);
+                }
+
+                if (_mode == Mode.Data)
+                {
+                    ReadChunkedData(readableBuffer, writableBuffer, out consumed, out examined);
+
+                    if (_mode == Mode.Data)
+                    {
+                        return false;
+                    }
+
+                    readableBuffer = readableBuffer.Slice(consumed);
+                }
+
+                if (_mode == Mode.Suffix)
+                {
+                    ParseChunkedSuffix(readableBuffer, out consumed, out examined);
+
+                    if (_mode == Mode.Suffix)
+                    {
+                        return false;
+                    }
+
+                    readableBuffer = readableBuffer.Slice(consumed);
+                }
+            }
+
+            // Chunks finished, parse trailers
+            if (_mode == Mode.Trailer)
+            {
+                ParseChunkedTrailer(readableBuffer, out consumed, out examined);
+
+                if (_mode == Mode.Trailer)
+                {
+                    return false;
+                }
+
+                readableBuffer = readableBuffer.Slice(consumed);
+            }
+
+            // _consumedBytes aren't tracked for trailer headers, since headers have separate limits.
+            if (_mode == Mode.TrailerHeaders)
+            {
+                if (_context.TakeMessageHeaders(readableBuffer, out consumed, out examined))
```

Comment:
We really need to stop putting trailers into the normal header collection.

Reply:
#4771 is currently in the backlog